### DISCUSSION TITLE
Add to Node.js CITGM

### DISCUSSION THREAD

**Discussion by GeoffreyBooth** [pnpm/pnpm#7239] [General]
Hi, in https://github.com/nodejs/citgm/pull/1020 I’m adding support for `pnpm` to Node’s Canary in the Gold Mine project, which aims to test new builds of Node.js against many popular community projects to inform Node maintainers of potential breakages before new versions of Node are released. I’m adding support both for `pnpm` to be used as one of the package managers (like how CITGM already supports `npm` and Yarn) and also as one of the libraries that get tested. Having `pnpm` here will help ensure that new versions of Node don’t break `pnpm`, and having support for `pnpm` as one of the package managers in CITGM means that we can add other projects to CITGM that use `pnpm` as their package manager.

Can someone please help me determine the best way to build and test `pnpm` in this environment? In https://github.com/nodejs/citgm/pull/1020/commits/46781754cff5188828df3a1a957669fac92cb470 I added the commands `pnpm run compile` and `pnpm run test-main` to build and test `pnpm`, but it fails in CI: https://github.com/nodejs/citgm/actions/runs/6612820965/job/17959419464. The `pnpm: true` part tells CITGM to use `pnpm` to install packages and to run these build and test commands. I’m not familiar with the `pnpm` codebase itself; this was based on looking in https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md and https://github.com/pnpm/pnpm/blob/main/.github/workflows/ci.yml. Any advice is greatly appreciated!

**Comment by zkochan** [comment]
At first glance everything looks correct. Although, it is weird that tests of only 3 workspace projects were started.

**Reply by GeoffreyBooth** [reply]
> it is weird that tests of only 3 workspace projects were started.

I included `PNPM_WORKERS=3` per https://github.com/pnpm/pnpm/blob/b90029432d66b24b2453e1755051bc782eaf98c3/.github/workflows/ci.yml#L70, perhaps that explains it?

On my local in Node 21.1.0, when I run `git clone git@github.com:pnpm/pnpm.git; cd pnpm; pnpm install; pnpm run compile-only; pnpm run test-main` I get one failing test, `env add/remove › install and remove multiple Node.js versions in one command` (it just times out). Are these the commands I should be running to test pnpm?

Looking at https://github.com/nodejs/citgm/actions/runs/6629821509/job/18009944592, I see

```
info: pnpm pnpm:          | test suite started  
error: pnpm pnpm:          | pnpm-test Timed Out 
error: failure             | Test Timed Out      
error: failing module(s)   |                     
error: module name:        | pnpm                
error: version:            | 8.9.2               
error: error:              | Test Timed Out      
error: error:              | undefinedScope: all [14](https://github.com/nodejs/citgm/actions/runs/6629821509/job/18009944592#step:5:15)4 workspace projects                                                                                                                                                                                                                                                                                                                                                                                                                                
error:                     | .../ghooks@2.0.4/node_modules/ghooks install$ node ./bin/module-install                                                                                                                                                                                                                                                                                                                                                                                                   
error:                     | .../ghooks@2.0.4/node_modules/ghooks install: This does not seem to be a git project.                                                                                                                                                                                                                                                                                                                                                                                     
error:                     | .../ghooks@2.0.4/node_modules/ghooks install: Although ghooks was installed, the actual git hooks have not.                                                                                                                                                                                                                                                                                                                                                               
error:                     | .../ghooks@2.0.4/node_modules/ghooks install: Run "git init" and then "npm explore ghooks -- npm run install".                                                                                                                                                                                                                                                                                                                                                            
error:                     | .../ghooks@2.0.4/node_modules/ghooks install: Please ignore this message if you are not using ghooks directly.                                                                                                                                                                                                                                                                                                                                                            
error:                     | .../ghooks@2.0.4/node_modules/ghooks install: Done                                                                                                                                                                                                                                                                                                                                                                                                                        
error:                     | .../esbuild@0.[19](https://github.com/nodejs/citgm/actions/runs/6629821509/job/18009944592#step:5:20).4/node_modules/esbuild postinstall$ node install.js                                                                                                                                                                                                                                                                                                                                                                                                      
error:                     | .../esbuild@0.19.4/node_modules/esbuild postinstall: Done                                                                                                                                                                                                                                                                                                                                                                                                                 
error:                     | .../node_modules/fuse-native install$ node-gyp-build                                                                                                                                                                                                                                                                                                                                                                                                                      
error:                     | .../node_modules/fuse-native install: make: Entering directory '/tmp/035[29](https://github.com/nodejs/citgm/actions/runs/6629821509/job/18009944592#step:5:30)a3c-6df5-4cac-a4b3-58f7897b6352/pnpm/node_modules/.pnpm/fuse-native@2.2.6/node_modules/fuse-native/build'                                                                                                                                                                                                                                                                                       
error:                     | .../node_modules/fuse-native install:   CC(target) Release/obj.target/fuse/fuse-native.o                                                                                                                                                                                                                                                                                                                                                                                  
error:                     | .../node_modules/fuse-native install:   SOLINK_MODULE(target) Release/obj.target/fuse.node                                                                                                                                                                                                                                                                                                                                                                                
error:                     | .../node_modules/fuse-native install:   COPY Release/fuse.node                                                                                                                                                                                                                                                                                                                                                                                                            
error:                     | .../node_modules/fuse-native install:   COPY ../build/Release/libfuse.so                                                                                                                                                                                                                                                                                                                                                                                                  
error:                     | .../node_modules/fuse-native install:   TOUCH Release/obj.target/postinstall.stamp                                                                                                                                                                                                                                                                                                                                                                                        
error:                     | .../node_modules/fuse-native install: make: Leaving directory '/tmp/0[35](https://github.com/nodejs/citgm/actions/runs/6629821509/job/18009944592#step:5:36)29a3c-6df5-4cac-a4b3-58f7897b6352/pnpm/node_modules/.pnpm/fuse-native@2.2.6/node_modules/fuse-native/build'                                                                                                                                                                                                                                                                                        
error:                     | .../node_modules/fuse-native install: Done
```

Perhaps something about how pnpm is checked out in CITGM might be affecting it? A shallow clone, no git hooks running, something along those lines? Are there things in the environment that pnpm’s tests are expecting?

**Reply by zkochan** [reply]
So the timeout issue might be real. We probably need to use some mocking as the server that hosts the node.js artifacts is super slow and makes the tests time out.

**Reply by GeoffreyBooth** [reply]
Well I don't need the tests to pass on my local, I only care about getting them to pass in the CITGM CI. Any thoughts about that? Am I at least running the correct setup and test commands?

**Reply by zkochan** [reply]
I think yes. You run the correct commands.

**Reply by GeoffreyBooth** [reply]
Then do you have any idea why the tests are failing? Are there any requirements of the environment that might not be being met?

**Reply by nachoaldamav** [reply]
I tested with more workers and the result is the same, some tests fail with timeouts (or randomly exiting with error code 1) and I think most of the times it fails with the `pnpm recursive exec works with PnP` test. This test also fails for me locally.

One thing I noticed is that the test runs `compile-only` and `test-main`, right? I think you can remove the `compile-only` as it already runs it when using `test-main`.

**Reply by GeoffreyBooth** [reply]
> So the timeout issue might be real. We probably need to use some mocking as the server that hosts the node.js artifacts is super slow and makes the tests time out.

I adjusted how the tests are run to ensure that the full git repo is checked out, and now the tests are at least running if not yet passing. Is there a way to run "core" tests, like things that always should pass and don't have dependencies on network downloads or other external and potentially flaky things?

**Reply by GeoffreyBooth** [reply]
On my local Mac with latest main, I see consistent failures with `pnpm recursive exec works with PnP`.

How would you feel if there was some way to run the pnpm test suite with flaky tests excluded? Like maybe a certain environment variable, when set, would exclude particular tests that aren’t guaranteed to pass. Then we could add pnpm to CITGM and run its tests with this flag, and we’d know that if the tests failed then Node somehow broke pnpm.

It’s a pretty hard rule of CITGM that we can’t add modules to test that don’t have rock solid test suites that are without flakes, because we need to avoid false positives where modules’ test suites fail due to factors other than the new Node code.

**Reply by zkochan** [reply]
There are only 2-3 tests that are "flaky" (which I guess we can either fix or skip on CITGM).

There are a few other tests that fail with timeout, when they try to download the node.js tarball. These we need to probably address somehow by mocking the node.js tarball request.

And there are some tests that sometimes fail but always pass on second run. Hence, we tell jest setup to retry a test 1 time if it fails.

Testing the "@pnpm/core" library might be enough. It should cover most of the things.

**Reply by GeoffreyBooth** [reply]
Can you tell me a command to run that skips the flaky tests, and also skips any tests that depend on external resources like downloads?

> Testing the “@pnpm/core” library might be enough.

What would be the command to do this?

**Reply by zkochan** [reply]
> Can you tell me a command to run that skips the flaky tests, and also skips any tests that depend on external resources like downloads?

There's no such command. Almost all tests make network requests to a locally running verdaccio. But those are stable. The unstable ones are only those that download and install Node.js.

>> Testing the “@pnpm/core” library might be enough.
> 
>What would be the command to do this?

```
pnpm -F core test
```

**Reply by GeoffreyBooth** [reply]
> Almost all tests make network requests to a locally running verdaccio. But those are stable. The unstable ones are only those that download and install Node.js.

Network requests to localhost don’t concern me; that should never fail in CITGM’s CI. It’s the tests that rely on external resources like the Node.js binary download server that I’d like to disable. How would you feel about adding an environment variable or other method that would allow me to skip such tests?

**Comment by GeoffreyBooth** [comment]
If you look at https://github.com/nodejs/citgm/actions/runs/6748209270, I have the `pnpm` tests passing in CITGM for a few of the test platforms and Node versions. Do you mind helping me understand the failures, and decide what you think we should do about them?

- The three Windows runs all failed on `Error: fuse-shared-library is not currently supported on: win32`. Do you expect your test suite to pass on Windows? If not, perhaps certain tests could be marked to be skipped on Windows?

- The macOS tests both fail on ``ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @pnpm/core@13.1.5 test: `pnpm run compile && pnpm run _test` ``. What does this mean? And for some reason the [macOS Node 20 run](https://github.com/nodejs/citgm/actions/runs/6748209270/job/18346068328) doesn’t print the [summary line](https://github.com/nodejs/citgm/actions/runs/6748209270/job/18346068687#step:5:2368) with the number of tests passed and failed.

**Reply by zkochan** [reply]
> The three Windows runs all failed on Error: fuse-shared-library is not currently supported on: win32. Do you expect your test suite to pass on Windows? If not, perhaps certain tests could be marked to be skipped on Windows?

That is not a test suite. That is a postinstall script of an optional dependency. It doesn't matter that it failed as it is an optional dependency. It is expected to fail on Windows.

**Reply by GeoffreyBooth** [reply]
Is it optional for building pnpm running its tests? Should I install pnpm without its optional dependencies?

**Reply by GeoffreyBooth** [reply]
It seems like `pnpm` can’t function without its optional dependencies:

```
.../esbuild@0.19.5/node_modules/esbuild postinstall$ node install.js                                                                                                                         
.../esbuild@0.19.5/node_modules/esbuild postinstall: [esbuild] Failed to find package "@esbuild/darwin-x64" on the file system                                                               
.../esbuild@0.19.5/node_modules/esbuild postinstall: This can happen if you use the "--no-optional" flag. The "optionalDependencies"                                                         
.../esbuild@0.19.5/node_modules/esbuild postinstall: package.json feature is used by esbuild to install the correct binary executable                                                        
.../esbuild@0.19.5/node_modules/esbuild postinstall: for your current platform. This install script will now attempt to work around                                                          
.../esbuild@0.19.5/node_modules/esbuild postinstall: this. If that fails, you need to remove the "--no-optional" flag to use esbuild.                                                        
.../esbuild@0.19.5/node_modules/esbuild postinstall: [esbuild] Trying to install package "@esbuild/darwin-x64" using npm                                                                     
.../esbuild@0.19.5/node_modules/esbuild postinstall: Done
```

So does this just mean that the test suite can’t run in Windows? Do you not test `pnpm` in Windows?

And can you explain the error I’m seeing in the Mac runs?

**Reply by zkochan** [reply]
Failing optional dependencies don't make installation fail with a non-0 exit code. Installation works on Windows. Fuse is mocked in tests.

**Reply by zkochan** [reply]
On macOS this two tests failed:

```
error:                     | FAIL test/install/misc.ts (414.172 s)                                                                                                               
error:                     | ● building native addons                                                                                                                            
error:                     |                                                                                                                                                     
error:                     | diskusage@1.1.3 install: `node-gyp rebuild`                                                                                                         
error:                     | Exit status 1                                                                                                                                       
error:                     |                                                                                                                                                     
error:                     | at EventEmitter.<anonymous> (../../node_modules/.pnpm/@pnpm+npm-lifecycle@2.0.2_typanion@3.14.0/node_modules/@pnpm/npm-lifecycle/index.js:286:13)   
error:                     | at ChildProcess.<anonymous> (../../node_modules/.pnpm/@pnpm+npm-lifecycle@2.0.2_typanion@3.14.0/node_modules/@pnpm/npm-lifecycle/lib/spawn.js:55:14)
error:                     |                                                                                                                                                     
error:                     | FAIL test/install/lifecycleScripts.ts (30.123 s)                                                                                                    
error:                     | ● lifecycle scripts have access to node-gyp                                                                                                         
error:                     |                                                                                                                                                     
error:                     | drivelist@5.1.8 install: `node-gyp rebuild`                                                                                                         
error:                     | Exit status 1                                                                                                                                       
error:                     |                                                                                                                                                     
error:                     | at EventEmitter.<anonymous> (../../node_modules/.pnpm/@pnpm+npm-lifecycle@2.0.2_typanion@3.14.0/node_modules/@pnpm/npm-lifecycle/index.js:286:13)   
error:                     | at ChildProcess.<anonymous> (../../node_modules/.pnpm/@pnpm+npm-lifecycle@2.0.2_typanion@3.14.0/node_modules/@pnpm/npm-lifecycle/lib/spawn.js:55:14)
```

We can check if we can rewrite them or skip them.

**Reply by GeoffreyBooth** [reply]
I looked at the Windows runs again and you’re right, it’s not failing on the install step; it runs the test suite and for Node 20 and 21, there are [some failing tests](https://github.com/nodejs/citgm/actions/runs/6748209270/job/18346070323#step:5:2373):

```
error:                     | Summary of all failing tests                                                                                                                                    
error:                     | FAIL test/install/peerDependencies.ts (28.499 s)                                                                                                                
error:                     | ● run pre/postinstall scripts of each variations of packages with peer dependencies                                                                             
error:                     |                                                                                                                                                                 
error:                     | @pnpm.e2e/pkg-with-events-and-peers@1.0.0 preinstall: `hello-world-js-bin && node create generated-by-preinstall`                                               
error:                     | spawn C:\Windows\system32\cmd.exe ENOENT                                                                                                                        
error:                     |                                                                                                                                                                 
error:                     |                                                                                                                                                                 
error:                     |                                                                                                                                                                 
error:                     | FAIL test/install/lifecycleScripts.ts (17.934 s)                                                                                                                
error:                     | ● scripts have access to unlisted bins when hoisting is used                                                                                                    
error:                     |                                                                                                                                                                 
error:                     | @pnpm.e2e/pkg-that-calls-unlisted-dep-in-hooks@1.0.0 postinstall: `node -e "process.stdout.write('Hello world!')" | json-append output.json`                    
error:                     | spawn C:\Windows\system32\cmd.exe ENOENT
```

For Node 18, the test suite [times out](https://github.com/nodejs/citgm/actions/runs/6748209270/job/18346069803#step:5:953).

**Comment by GeoffreyBooth** [comment]
I’ve opened a PR at https://github.com/nodejs/citgm/pull/1027 specifically for adding `pnpm` to the Node CITGM test suite. Once you think `pnpm`‘s test suite is stable enough that it should pass consistently in the CITGM GitHub Actions runs, please comment on that PR and we can test it.

I would like to include `pnpm` in Node CITGM but I don’t have any more time to spend on this; if you would like it in CITGM, so that future Node updates hopefully don’t break `pnpm`, then I’ll need someone from the `pnpm` project to take it from here and get a test suite that is stable and passes reliably and remains so.

**Reply by zkochan** [reply]
Thanks for the effort. We'll look into it but we'll need some time to get back to it.
